src/osx/cocoa/textctrl.mm

```diff
@@ -1142,80 +1367,256 @@ - (void)controlTextDidEndEditing:(NSNotification *)aNotification
+                    [paraStyle setHyphenationFactor:0.0];
+                else
+                    [paraStyle setHyphenationFactor:1.0];
+            }
+
+            if ( style.HasTabs() )
+            {
+                while ([[paraStyle tabStops] count] > 0)
+                {
+                    NSTextTab* tab = [[paraStyle tabStops] objectAtIndex:0];
+                    [paraStyle removeTabStop:tab];
+                }
+
+                wxArrayInt tabs = style.GetTabs();
+                for (int i = 0; i < tabs.size(); i++)
+                {
+                    long position = tabs[i];
+                    NSTextTab* tab = [[NSTextTab alloc]
+                                      initWithType:NSLeftTabStopType
+                                      location:PointsFromInternalUnits(position, style.GetUnitsPerPoint())];
+
+                    [paraStyle addTabStop:tab];
+                }
             }
-            [dict setObject:[NSNumber numberWithInt:( underlineStyle )] forKey:NSUnderlineStyleAttributeName];
-            wxColour colour = style.GetUnderlineColour();
-            if ( colour.IsOk() )
+
+            [storage removeAttribute:NSParagraphStyleAttributeName range:range];
+            [storage addAttribute:NSParagraphStyleAttributeName value:paraStyle range:range];
+            [paraStyle release];
+
+            if ( style.HasParagraphUserData() )
             {
-                [dict setValue:colour.OSXGetNSColor() forKey:NSUnderlineColorAttributeName];
+                [storage removeAttribute:wxNSParagraphUserDataAttributeName range:range];
+                [storage addAttribute:wxNSParagraphUserDataAttributeName value:[NSNumber numberWithLong:style.GetParagraphUserData()] range:range];
             }
-            [storage addAttributes:dict range:range];
-            [dict release];
         }
     }
+    @catch (NSException *exception)
+    {
+        NSLog(@"%@", exception);
+    }
 
-    if ( style.HasAlignment() )
+    if ( start != end )
+        [m_textView didChangeText]; // for Undo
+
+    // This puts the style change operation as its own item in the Undo stack.
+    if ( start != end && !style.IsParagraphStyle() && ![m_textView isCoalescingUndo] && CanUndo() )
     {
-        switch ( style.GetAlignment() )
-        {
-            case wxTEXT_ALIGNMENT_RIGHT:
-                [m_textView setAlignment:NSRightTextAlignment];
-                break;
-            case wxTEXT_ALIGNMENT_CENTER:
-                [m_textView setAlignment:NSCenterTextAlignment];
-                break;
-            default:
-                [m_textView setAlignment:NSLeftTextAlignment];
-                break;
-        }
+        Undo();
+        Redo();
     }
 }
 
 void wxNSTextViewControl::CheckSpelling(bool check)
 {
     if (m_textView)
+    {
         [m_textView setContinuousSpellCheckingEnabled: check];
+        [m_textView setAutomaticSpellingCorrectionEnabled: check];
```

Comment:
Hmm, I wonder why hadn't this been done when adding this code initially. Can you think of any reason why we would _not_ want to do this?

Reply:
because autocorrection quite frequently can be false, at least for me when writing in different languages or dialects, that way you just get a hint, if you see the spellchecker is right you can let it correct, but otherwise if it's done automatically I end up writing the same correct word several times, fighting the spell checker